src/Streamly/Internal/Data/Unfold/Enumeration.hs

```diff
@@ -0,0 +1,336 @@
+-- [1.1,2.1,3.1,4.1]
+--
+-- @
+--
+-- /Pre-release/
+--
+    enumerateFrom :: (Integral a, Monad m) => Unfold m a a
+
+-- Generate a finite stream starting with the element @from@, enumerating
+-- the type up to the value @to@. If @to@ is smaller than @from@ then an
+-- empty stream is returned.
+--
+-- @
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4) 0
+-- [0,1,2,3,4]
+--
+-- @
+--
+-- For 'Fractional' types, the last element is equal to the specified @to@
+-- value after rounding to the nearest integral value.
+--
+-- @
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4) 1.1
+-- [1.1,2.1,3.1,4.1]
+--
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromTo 4.6) 1.1
+-- [1.1,2.1,3.1,4.1,5.1]
+--
+-- @
+--
+-- /Pre-release/
+    enumerateFromTo :: (Integral a, Monad m) => a -> Unfold m a a
+
+-- @enumerateFromThen from then@ generates a stream whose first element
+-- is @from@ and the successive elements are
+-- in increments of @then@.  Enumeration can occur downwards or
+-- upwards depending on whether @then@ comes before or after @from@. For
+-- 'Bounded' types the stream ends when 'maxBound' is reached, for
+-- unbounded types it keeps enumerating infinitely.
+--
+-- @
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold (Unfold.enumerateFromThen 2) 0
+-- [0,2,4,6]
+--
+-- >>> Stream.toList $ Stream.take 4 $ Stream.unfold (Unfold.enumerateFromThen (-2)) 0
+-- [0,-2,-4,-6]
+--
+-- @
+--
+-- /Pre-release/
+    enumerateFromThen :: (Integral a, Monad m) => a -> Unfold m a a
+
+-- @enumerateFromThenTo from then to@ generates a finite stream whose
+-- first element is @from@ and the successive elements are in
+-- increments of @then@ up to @to@. Enumeration can
+-- occur downwards or upwards depending on whether @then@ comes before or
+-- after @from@.
+--
+-- @
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromThenTo 2 6) 0
+-- [0,2,4,6]
+--
+-- >>> Stream.toList $ Stream.unfold (Unfold.enumerateFromThenTo (-2) (-6)) 0
+-- [0,-2,-4,-6]
+--
+-- @
+--
+-- /Pre-release/
+    enumerateFromThenTo :: (Integral a, Monad m) => a-> a -> Unfold m a a
+
+-------------------------------------------------------------------------------
+-- Enumerable Instances
+-------------------------------------------------------------------------------
+--
+-- For Enum types smaller than or equal to Int size.
+#define ENUMERABLE_BOUNDED_SMALL(SMALL_TYPE)           \
+instance Enumerable SMALL_TYPE where {                 \
+    {-# INLINE enumerateFrom #-};                      \
+    enumerateFrom = enumerateFromIntegral;             \
+    {-# INLINE enumerateFromTo #-};                    \
+    enumerateFromTo = enumerateFromToIntegral;         \
+    {-# INLINE enumerateFromThen #-};                  \
+    enumerateFromThen = enumerateFromThenIntegral;     \
+    {-# INLINE enumerateFromThenTo #-};                \
+    enumerateFromThenTo = enumerateFromThenToIntegral }
```

Comment:
I am wondering how is this compiling, these types are not Integral types. How can we define them using enumerateFromIntegral etc?